### PR TITLE
MySQL support for `bbs_database` + integration test

### LINKED ISSUE #431: MySQL support in database cration
## 🚀 Feature
Add MySQL support for database creation. One should be able to 
```
bbs_database init --db-type mysql ...
```
```
bbs_database add --db-type mysql ...
```
They both currently raise `NotImplementedError`.
## Motivation
#412 introduced a first version of the database creation logic. To keep things simple, the only supported backend was SQLite. However, to be able to host the database on a remote server we should provide MySQL support.

## Pitch
We are currently using MySQL for the Kaggle database and it works pretty well.

## Alternatives
Note that there is an issue about migrating to MariaDB #416

## Discussion
SQLAlchemy abstracts out the backend in theory. However, we saw on multiple occasions that the behavior differs. Therefore it is still important to write tests (+ maybe some setup logic). We can take inspiration from what was done with the Kaggle database. See below the relevant parts of the codebase
* [mysql Dockerfile](https://github.com/BlueBrain/Search/blob/a37bfca303728e3a8cfe94abff533584b44e04ea/docker/mysql.Dockerfile)
* [Testing it locally](https://github.com/BlueBrain/Search/blob/a37bfca303728e3a8cfe94abff533584b44e04ea/tests/conftest.py#L232-L299) (maybe there is a simpler solution?)

### PR DESCRIPTION
**Pull request by jankrepl**
Closes #431 

# What was done
* Made sure `bbs_database init` and `bbs_database add` support `--db-type=mysql`
* Introduced 2 new subfolders inside of `tests` (inspired by https://realpython.com/python-testing/#writing-integration-tests)
  * `unit`
  * `integration`
* Wrote `tests/integration/test_bbs_database.py` integration test that tests all the existing commands (`init`, `parse` and `add`) without any mocking/patching and using both `sqlite` and `mysql` backends

# How to test
```bash
pytest tests/integration/test_bbs_database.py
```
This should run both the `sqlite` and `mysql` test. If the `mysql` test is skipped then you probably don't have docker running on your computer (check `docker ps`). Also note that you need to get the MySQL image `docker pull mysql`.

# More details + to discuss
### So the `bbs_database` unit tests are only run for the `sqlite` backend, right?
Yes. This is purely because of the fact that the mysql docker container just takes too long to start. We have this `sqlite` and `mysql` logic in the current codebase (for the CORD19 database) and a lot of us just skip the `mysql` by doing `pytest -k "not mysql"`.
### Should we modify `tox.ini` and have different handling of unit and integrations tests?
I guess this is something we should decide on together. In general, the integration tests should not do any patching/mocking and are gonna take longer to run.

### Isn't the `docker` +`mysql` logic just living at two places now (`tests/conftest.py` and `tests/integration/test_bbs_database.py`)?
Yes, and there is a big overlap. However, I propose we create a separate issue for refactoring things. For now, our test suit is going to create 2 docker containers.

### CODE REVIEW

**Comment by Stannislav** [comment]
I'll review the code later, just two thoughts I had looking at the changes:
- With the new structure of `tests/` we can actually move `benchmarks` to `tests/benchmarks` (new issue?)
- Maybe we cen split the tests in our CI/tox into `unit`/`integration`/`benchmarks` (new issue? need to think though the details...)

**Comment by jankrepl** [comment]
> * With the new structure of `tests/` we can actually move `benchmarks` to `tests/benchmarks` (new issue?)

Nice idea! This makes me think about the fact that we should probably refactor the `benchmarks` since they are not really used.

> * Maybe we cen split the tests in our CI/tox into `unit`/`integration`/`benchmarks` (new issue? need to think though the details...)

Yes! I mentioned something similar in the description. Related to this, I think we could also refactor our current unit tests and turn some of them into integration tests and make sure that unit tests are really unit tests.

**Review by pafonta — APPROVED** [review]
I haven't found the _How to test?_ section in the PR description.

So, I have tried the following but the added test for `mysql` is skipped.

How to not skip the test for `mysql`?

```
pytest tests/integration/test_bbs_database.py
```

File: tox.ini [hunk 1]
```diff
@@ -152,6 +152,7 @@ per-file-ignores =
 [isort]
 profile = black
 honor_noqa = true
+known_third_party = docker
```
[the remark this answers is not in the document]
Comment (jankrepl) [reply tox.ini:155]: It is because we have a folder `docker` in the root of our repo:D

**Comment by jankrepl** [comment]
> I see that the GitHub CI tests pass.
> 
> However, I haven't found the _How to test?_ section in the PR description. So, I have tried the following but the added test for `mysql` is skipped.
> 
> How to not skip the test for `mysql`?
> 
> ```
> pytest tests/integration/test_bbs_database.py
> ```

Great point. I modified the description:)

My best guess on why the `mysql` test was skipped on your computer is because you don't have the `docker` daemon running locally. Try running `docker ps` to check. Also, you will need to `docker pull mysql` to get the MySQL image.

Here is the code that checks for the docker daemon
https://github.com/BlueBrain/Search/blob/a7687a38a62536478a2058da10938f2bc47e4d47/tests/integration/test_bbs_database.py#L19-L25

**Review by Stannislav — APPROVED** [review]
Nice, thanks :)

File: tests/integration/test_bbs_database.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+import docker
+import pytest
+import sqlalchemy
+from sqlalchemy.exc import OperationalError
+
+from bluesearch.entrypoint.database.parent import main
+
+
+@pytest.fixture(params=["sqlite", "mysql"])
+def setup_backend(request, tmpdir):
```
[the remark this answers is not in the document]
Comment (jankrepl) [reply tests/integration/test_bbs_database.py]: Done in b0841ac

**Pull request merged**